docs: Add content for new `strict` mode for Angular CLI (#37486)
angular/angular: aio/content/guide/strict-mode.md

# Strict mode

When you create a new workspace or a project you have an option to create them in a strict mode using the `--strict` flag.

> Enabling this flag initializes your new project with a few new settings that improve maintainability, help you catch bugs ahead of time, and allow the CLI to perform advanced optimizations on your app.

Reviewer: ```suggestion
Enabling this flag initializes your new workspace with a few new settings that improve maintainability, help you catch bugs ahead of time, and allow the CLI to perform advanced optimizations on your application.
```

Additionally, applications that use these stricter settings are easier to statically analyze, which can help the `ng update` command refactor code more safely and precisely when you are updating to future versions of Angular.

Specifically, the `strict` flag does the following:

* Enables [`strict` mode in TypeScript](https://www.staging-typescript.org/tsconfig#strict), as well as other strictness flags recommended by the TypeScript team. Specifically, `forceConsistentCasingInFileNames`, `noImplicitReturns`,  `noFallthroughCasesInSwitch`.
* Turns on strict Angular compiler flags [`strictTemplates`](guide/angular-compiler-options#stricttemplates) and [`strictInjectionParameters`](guide/angular-compiler-options#strictinjectionparameters)
* [Bundle size budgets](guide/build#configuring-size-budgets) have been reduced by ~75%
* Turns on [`no-any` tslint rule](https://palantir.github.io/tslint/rules/no-any/) to prevent declarations of type `any`
* [Marks your application as side-effect free](https://webpack.js.org/guides/tree-shaking/#mark-the-file-as-side-effect-free) to enable more advanced tree-shaking

> You can apply these settings at the workspace and project level.

Reviewer: Umm.. not quite right, there are strict mode settings which are workspace wide and strict mode settings which are at project level. Some settings are workspace level and some are at application project level.

I think it would also be good to clarify this in the docs. As using `ng generate application [project-name] --strict` in a non strict workspace is different from `ng generate application [project-name] --strict` in a strict workspace.

See 
https://github.com/angular/angular/pull/37486#issuecomment-644600427 for more info.

Reviewer 2: @alan-agius4 I've just tested `ng generate application [project-name] --strict` in a non-strict workspace and it doesn't seem to generate a strict project.

In general the diff between strict application in a non-strict workspace and an application within a strict workspace should be minimal or none. Anything else would likely be perceived as bug. Can you please look into why this doesn't seem to work as expected? thanks

Reviewer 2: For the time being, let's resolve this conversation and keep the documentation as is until we have more information about why this must work differently.

Reviewer: ## The Why
It works differently because `strict` on `ng new` is a combination of 2 schematics. the workspace and the application schematics. Both of which set different strict settings. 

The `ng new` schematic is just a "wrapper" which combines both of these into a single schematic which generates a workspace and an application. This is equivalent in running `ng g workspace myapp --strict/--no-strict` and `ng g app myapp --strict/--no-strict`. 
 
But when you use `ng g app --strict` you will only get the strict settings which are available/set only at application level, such as bundle budgets, side effects package.json etc.., you will not get the workspace strict settings such as stricter TSC compiler options, stricter tslint rules etc...

## The What
### ng-new/workspace schematic

- We add stricter TypeScript and Angular type-checking options which effect all the projects in a workspace. (These are the settings that you mentioned in this PR).
- Add stricter tslint rules ex: no-any which effect all the projects in a workspace.
- When createApplication is true, which is the default, the generated root application is also created with strict settings (described below).

### application schematic

- A package.json is added in the app folder, to tell the bundlers whether the application is free from side-effect code. When strict is true. the sideEffects will be set false.
- Stricter application bundle budgets.

## Epilogue
Thinking about this, maybe we should hide the `strict` option from the application schematic, and only generate strict mode applications in strict workspaces.

**Note**: future applications are always generated in strict mode in strict workspaces even when not using `--strict`.

To create a new workspace and application using the strict mode, run the following command:

<code-example language="sh" class="code-shell">

ng new [project-name] --strict

</code-example>

> To use strict mode for a new application project in an existing workspace, run the following command:

Reviewer 2: ```suggestion
To create a new project in the strict mode within an existing non-strict workspace workspace, run the following command:
```

> To create a new project in the strict mode within an existing non-strict workspace, run the following command:

Reviewer: ```suggestion
To create a new application in the strict mode within an existing non-strict workspace, run the following command:
```

`project` is a generic term which can mean a number of project types. Ex: Application and Library

<code-example language="sh" class="code-shell">

ng generate application [project-name] --strict

</code-example>

{@a side-effect}

> ### Determine if an application has non-local side effects

Reviewer 2: ```suggestion
### Configuring if an application has non-local side effects
```

# Employing strict mode when building applications

When you create a new workspace or application, you have the option to use a more strict project setup when you create a new workspace with `ng new`.

```
ng new --strict
```

> Enabling this flag initializes your new project with a few new settings that improve maintainability, help you catch bugs ahead of time, and allow the CLI to perform advanced optimizations on your app.

Reviewer: Can we avoid abbreviates. We should use application instead of app

In addition, using this strict mode can improves the reliability of updating your application using `ng update`.
Applications that use these stricter settings are easier to statically analyze, which can help the `ng update` command refactor code more safely and precisely.
Specifically, the `strict` flag does the following:

* Enables strict mode in TypeScript
* Turns template type checking to Strict
* Default bundle budgets have been reduced by ~75%
* Configures linting rules to prevent declarations of type `any`
* Configures your app as side-effect free to enable more advanced tree-shaking

You can apply these settings at the workspace and project level.

To create a new workspace and application using the strict mode, run the following command:

<code-example language="sh" class="code-shell">

ng new [project-name] --strict

</code-example>

> To use strict mode for a new project in an existing application, run the following command:

Reviewer: To use strict mode for a new application project in an existing workspace
